src/relay/ir/doc.h

```diff
@@ -0,0 +1,436 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file doc.h
+ * \brief A pretty printer DSL for constructing (Doc) and formatting (RDoc) documents.
+ *        It is based heavily on Philip Wadler's "A prettier printer."
+ *        See https://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf
+ *        for more details.
+ *
+ * Since the original paper uses call by value for efficiency, everything doc function is maximally lazy.
+ * You can probably yank speed by doing strict analysis and removing some Lazy (if this is bottleneck).
+ */
+#ifndef TVM_RELAY_IR_DOC_H_
+#define TVM_RELAY_IR_DOC_H_
+
+#include <map>
+#include <utility>
+#include <string>
+#include <functional>
+#include <vector>
+#include <memory>
+#include <ostream>
+#include "error.h"
+
+namespace tvm {
+namespace relay {
+
+struct DocError : public Error {
+  explicit DocError(const std::string& msg) : Error(msg) { }
+};
+
+template<typename T>
+struct LazyNode {
+  mutable std::function<T()> thunk;
+  explicit LazyNode(const std::function<T()>& thunk) : thunk(thunk) { }
+};
+
+//! \brief denote a value that will be computed (at most once) on need.
+template<typename T>
+struct Lazy {
+  std::shared_ptr<LazyNode<T>> ln;
```

Comment:
add space between ```> >``` to make sure compatible with old c++ compiler

Reply:
done